python/dgl/transform.py

```diff
@@ -170,21 +170,17 @@ def _knn_graph_topk(x, k):
         x = F.unsqueeze(x, 0)
     n_samples, n_points, _ = F.shape(x)
 
+    ctx = F.context(x)
     dist = pairwise_squared_distance(x)
     k_indices = F.argtopk(dist, k, 2, descending=False)
-    dst = F.copy_to(k_indices, F.cpu())
-
-    src = F.zeros_like(dst) + F.reshape(F.arange(0, n_points), (1, -1, 1))
-
-    per_sample_offset = F.reshape(F.arange(0, n_samples) * n_points, (-1, 1, 1))
-    dst += per_sample_offset
-    src += per_sample_offset
-    dst = F.reshape(dst, (-1,))
-    src = F.reshape(src, (-1,))
-    adj = sparse.csr_matrix(
-        (F.asnumpy(F.zeros_like(dst) + 1), (F.asnumpy(dst), F.asnumpy(src))),
-        shape=(n_samples * n_points, n_samples * n_points))
-    return convert.from_scipy(adj)
+    # index offset for each sample
+    offset = F.arange(0, n_samples, ctx=ctx) * n_points
+    offset = F.unsqueeze(offset, 1)
+    src = F.reshape(k_indices, (n_samples, n_points * k))
+    src = F.unsqueeze(src, 0) + offset
+    dst = F.repeat(F.arange(0, n_points, ctx=ctx), k, dim=0)
```

Comment:
Will this cause larger memory consumption in the process?

Reply:
Overall, the src and dst tensors are the same shape as before, so memory should be the same (but now on GPU if the input is on GPU).
And now we don't have the potential memory overhead of `sparse.csr_matrix` (I don't know if it was significant).